skrifa/src/meta/charmap.rs

```diff
@@ -0,0 +1,415 @@
+}
+
+#[derive(Clone)]
+struct Map<'a> {
+    subtable: UnicodeSubtable<'a>,
+    is_symbol: bool,
+}
+
+impl<'a> Map<'a> {
+    fn map(&self, codepoint: u32) -> Option<GlyphId> {
+        self.map_impl(codepoint).or_else(|| {
+            if self.is_symbol {
+                self.map_impl(Self::shift_symbol_to_pua(codepoint)?)
+            } else {
+                None
+            }
+        })
+    }
+
+    fn map_impl(&self, codepoint: u32) -> Option<GlyphId> {
+        match &self.subtable {
+            UnicodeSubtable::Format4(subtable) => subtable.map_codepoint(codepoint),
+            UnicodeSubtable::Format12(subtable) => subtable.map_codepoint(codepoint),
+        }
+    }
+
+    fn shift_symbol_to_pua(codepoint: u32) -> Option<u32> {
+        // From HarfBuzz:
+        // For symbol-encoded OpenType fonts, we duplicate the
+        // U+F000..F0FF range at U+0000..U+00FF.  That's what
+        // Windows seems to do, and that's hinted about at:
+        // https://docs.microsoft.com/en-us/typography/opentype/spec/recom
+        // under "Non-Standard (Symbol) Fonts".
+        // See <https://github.com/harfbuzz/harfbuzz/blob/453ded05392af38bba9f89587edce465e86ffa6b/src/hb-ot-cmap-table.hh#L1595>
+        if codepoint <= 0x00FF {
+            Some(codepoint + 0xF000)
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Clone)]
+enum UnicodeSubtable<'a> {
+    Format4(Cmap4<'a>),
+    Format12(Cmap12<'a>),
+}
+
+impl<'a> UnicodeSubtable<'a> {
+    fn new(subtable: CmapSubtable<'a>) -> Option<Self> {
+        Some(match subtable {
+            CmapSubtable::Format4(cmap4) => Self::Format4(cmap4),
+            CmapSubtable::Format12(cmap12) => Self::Format12(cmap12),
+            _ => return None,
+        })
+    }
+
+    fn from_cmap_record(cmap: &Cmap<'a>, record: &cmap::EncodingRecord) -> Option<Self> {
+        Self::new(record.subtable(cmap.offset_data()).ok()?)
+    }
+}
+
+/// The mapping kind of a cmap subtable.
+///
+/// The ordering is significant and determines the priority of subtable
+/// selection (greater is better).
+#[derive(Copy, Clone, PartialEq, PartialOrd)]
+enum MappingKind {
+    None = 0,
+    UnicodeBmp = 1,
+    UnicodeFull = 2,
+    Symbol = 3,
+}
+
+/// The result of searching the cmap table for the "best" available
+/// subtables.
+///
+/// For `map`, best means either symbol (which is preferred)
+/// or a Unicode subtable with the greatest coverage.
+///
+/// For `variant_map`, best means a format 14 subtable.
+struct MappingSelection<'a> {
+    /// The mapping index accelerator that holds indices of the following
+    /// subtables.
+    mapping_index: MappingIndex,
+    /// Either a symbol subtable or the Unicode subtable with the
+    /// greatest coverage.
+    map: Option<UnicodeSubtable<'a>>,
+    /// Subtable that supports mapping Unicode variation sequences.
+    variant_map: Option<Cmap14<'a>>,
+}
+
+impl<'a> MappingSelection<'a> {
```

Comment:
Thanks for moving the algorithm here, that makes it indeed easier to read.

Reply:
I agree. Thanks for the suggestion